### PR TITLE
Use dict for Index serialization

### PR DESCRIPTION
**Pull request by dbespalov**
@yurymalkov @dyashuni this PR is still a WIP and untested. I should be a able to finish it tomorrow

### CODE REVIEW

**Comment by dbespalov** [comment]
@yurymalkov PR is ready for the review. 

One item to potentially look into:  pybind11 complained when I tried returning dict as object's state, so I wrapped the parameter dict into tuple (of length one). I did not give it too much thought, tbh, so it might still be possible to avoid this.

File: python_bindings/bindings.cpp [hunk 1]
```diff
@@ -345,140 +348,153 @@ class Index {
...
+      appr_alg->maxlevel_ = d["max_level"].cast<int>();
+      appr_alg->enterpoint_node_ = d["enterpoint_node"].cast<hnswlib::tableint>();
 
-      appr_alg->maxlevel_ = t[6].cast<int>();
-      appr_alg->enterpoint_node_ = t[7].cast<hnswlib::tableint>();
+      assert_true(appr_alg->maxM_ == d["max_M"].cast<size_t>(), "Invalid value of maxM_ ");
+      assert_true(appr_alg->maxM0_ == d["max_M0"].cast<size_t>(), "Invalid value of maxM0_ ");
+      assert_true(appr_alg->M_ == d["M"].cast<size_t>(), "Invalid value of M_ ");
+      assert_true(appr_alg->mult_ == d["mult"].cast<double>(), "Invalid value of mult_ ");
+      assert_true(appr_alg->ef_construction_ == d["ef_construction"].cast<size_t>(), "Invalid value of ef_construction_ ");
```
Comment (yurymalkov) [inline python_bindings/bindings.cpp:486]: I think the messages might confuse the user. I would add something like "Unpickle error:...".

**Review by yurymalkov — APPROVED** [review]
@dbespalov Thanks for an awesome contribution! 
One cosmetic change I would suggest is to extend the assert messages, so in case of a failure a user would better understand what has happened.

File: python_bindings/bindings.cpp (hunk 1 above)
Comment (dbespalov) [reply python_bindings/bindings.cpp:486]: changed the error message to "Unpickle error"

**Comment by dbespalov** [comment]
@yurymalkov good point, changed the error message. Thank you!

**Comment by yurymalkov** [comment]
Great! Thank you so much!

**Pull request merged**